include/ppx/geometry.h

```diff
@@ -169,68 +169,64 @@ class Geometry
+            uint32_t writeOffset  = offset == -1 ? mOffset : offset;
 
-            if (mMaxElementCount == 0) {
-                // Allocate storage for incoming data
-                SetSize(mOffset + sizeOfValues);
+            if (offset == -1 && !mFinalSizeSet) {
+                // New data is being appended, need to allocate storage
+                SetSize(writeOffset + sizeOfValues);
+            }
+            else {
+                PPX_ASSERT_MSG(writeOffset + sizeOfValues <= GetSize(), "Attempting to write values of size " << sizeOfValues << " at offset " << offset << " will overflow the buffer of size " << GetSize());
             }
 
             // Copy data
             const void* pSrc = pValues;
-            void*       pDst = mData.data() + mOffset;
+            void*       pDst = mData.data() + writeOffset;
             memcpy(pDst, pSrc, sizeOfValues);
 
-            mOffset += sizeOfValues;
+            if (offset == -1) {
+                mOffset += sizeOfValues;
+            }
         }
 
-        // If Overwrite is used to write new data, calling code will also
-        // have to SetOffset() appropriately
-        //
+        // Write() is used to write data at the specified offset
+        // If called with offset=-1, will write at the end and adjust mOffset accordingly
         template <typename T>
-        void Overwrite(const T& value, size_t offset)
+        void Write(const T& value, int32_t offset = -1)
         {
             uint32_t sizeOfValue = static_cast<uint32_t>(sizeof(T));
+            uint32_t writeOffset = offset == -1 ? mOffset : offset;
 
-            PPX_ASSERT_MSG(offset + sizeOfValue <= GetSize(), "Attempting to overwrite with value of size " << sizeOfValue << " at offset " << offset << " will overflow the buffer of size " << GetSize());
+            if (offset == -1 && !mFinalSizeSet) {
+                // New data is being appended, need to allocate storage
+                SetSize(writeOffset + sizeOfValue);
+            }
+            else {
+                PPX_ASSERT_MSG(writeOffset + sizeOfValue <= GetSize(), "Attempting to write value of size " << sizeOfValue << " at offset " << offset << " will overflow the buffer of size " << GetSize());
+            }
 
             // Copy data
             const void* pSrc = &value;
-            void*       pDst = mData.data() + offset;
+            void*       pDst = mData.data() + writeOffset;
             memcpy(pDst, pSrc, sizeOfValue);
+
+            if (offset == -1) {
+                mOffset += sizeOfValue;
+            }
         }
 
     private:
-        BufferType        mType            = BUFFER_TYPE_VERTEX;
-        uint32_t          mElementSize     = 0;
-        uint32_t          mMaxElementCount = 0;
+        BufferType        mType         = BUFFER_TYPE_VERTEX;
+        uint32_t          mElementSize  = 0;
+        bool              mFinalSizeSet = false;
```

Comment:
- if this value is only set in constructor, you can put `const` here
- i feel `mUseFixedSize` might be better in this case. When I see `mFinalSizeSet`, i would assume the `final size` is set but might be changed. The goal here is to reduce confusion for people who are not familiar with the class, but i am open to discussions :)